# Week 8 Computer Tutorial
######  Dr Richi Nayak, r.nayak@qut.edu.au

**Tutorial Topics:**
1. [Association mining Introduction: Basics](#amib)
2. [Measures of interestingness of Association rules](#moiar)
3. [Apriori Algorithm](#ap)

**Practical Topics:**
1. [Performing Association Mining](#am)
2. [Performing Sequential Rule Mining Using SPMF](#sm)


# Part 1 - Reflective exercises
In the past weeks, you have learned the data exploration using summary statistics, visualization and data preparation. It helps to understand and prepare the dataset for the data mining models as shown in the following diagram. In this practical, you will be introduced to a new dataset, data preparation for association mining, and association and sequnce rule mining. Please reflect on the association mining topic and answer the questions provided in each subsection.

<img src="ProcessFlow.png" width=700 height=400 />

## Exercise 1: Association mining Introduction: Basics<a name="amib"></a>


1. A data mining algorithm designed to discover frequently accessed items that occur in the same order.

        a. serial miner
        b. association rule miner
        c. sequence miner
        d. decision miner 

2. For the given dataset, identify how many items, itemset and transactions are there.
<img src="TransactionDB.png" width=300 height=100 />

3. Unlike traditional decision rules, association rules 

        a. allow the same variable to be an input attribute in one rule and an output attribute in another rule.    
        b. allow more than one input attribute in a single rule.
        c. require input attributes to take on numeric values. 
        d. require each rule to have exactly one categorical output attribute. 

4. This approach is best when we are interested in finding all possible interactions among a set of attributes.

        a. Decision tree
        b. Association rules
        c. K-Means algorithm
        d. Regression function

## Exercise 2: Measures of interestingness of Association rules<a name="moiar"></a>

1. From the given example association rules, identify the invalid ones and explain why. 

        a. A,B => E,C
        b. A => B,C
        c. A,B => B,C
        d. A,B => {}
        e. {} => C,E
    
2. Given the following association rules result, interpret what does it mean? 
Diaper, Milk→ Beer [support =5%, confidence=80%] 

3. Consider the following transactional data
<img src="TransactionDB1.png" width=300 height=100 />
Answer the following questions:

        a. Compute the support for itemsets {A}, {B}, {C}, and {A, B}
        b. Compute the confidence and lift for the rules {A, C} -> B and {B, C} -> A. 
         
4. Association rule `support` is defined as

        a. the percentage of instances that contain the antecendent conditional items listed in the association rule. 
        b. the percentage of instances that contain the consequent conditions listed in the association rule. 
        c. the percentage of instances that contain all items listed in the association rule. 
        d. the percentage of instances in the database that contain at least one of the antecendent conditional items listed in the association rule. 

5. Given a rule of the form IF X THEN Y, rule confidence is defined as the conditional probability that

        a. Y is true when X is known to be true.
        b. X is true when Y is known to be true.
        c. Y is false when X is known to be false.
        d. X is false when Y is known to be false.
    
6. Calculate the confidence of rules A → BCD, and ABC → D given their support?

7. Given a frequent itemset (ABCD), 
   - Generate all its frequent 3-itemset subsets.
   - Generate all the association rules with three items on LHS and one item on RHS?


## Exercise 3: Apriori Algorithm<a name="ap"></a>
1. Suppose you use the Apriori algorithm on the dataset with a minimum support threshold of 20%. How many candidate and frequent itemsets will be generated? Generate at least 10 association rules assuming that the confidence threshold is at 50%. 
<img src="TransactionDB2.png" width=300 height=100 />

2. If the minimum support threshold is zero, how many candidate itemsets must be generated by the algorithm?

3. How many candidate items are generated in the first scan of the algorithm?

4. Consider an example of a supermarket database which might have several million transactions and several thousand items of which only 1000 items are frequent. Which part of the Apriori algorithm will be most expensive to compute and Why?

5. Given this transactional data, identify the problems that will be faced with an association mining algorithm.

| Transaction Id |Items Bought|
| --- | --- |
| 001 |{A}|
| 002 |{A}|
| 003 |{B,C}|
| 004 |{D}|
| 005 |{A}|
| 006 |{D}|
| 007 |{A}|
| 008 |{A}|
| 009 |{A}|
| 010 |{B,B}|

# Part 2 - Practical exercises
___
A transactional dataset called `bank.csv` is given for you to apply association mining and sequence mining. 

## 1. Data Mining <a name="intro"></a>
After the data is cleaned, models can be built to perform in-depth analysis. There are two broad categories of data mining: **Predictive Mining** (e.g. classification and regression with decision tree, regression and neural network) and **Descriptive Mining** (e.g. clustering and association mining). Many algorithms belonging to each of the categories are available in `sklearn`, each with its own characteristics. The upcoming practical notes will explore some of these algorithms in detail.

Data mining outcomes are best understood when accompanied with graphs and charts of patterns and trends identified in the data. Visualisation allows us to understand the data better. In this unit, all visualisations will be done using `seaborn` and `matplotlib` with data presented by `pandas` DataFrames.

## 2. Performing Association Mining <a name="am"></a>

In these notes, we will focus on how to pre-process the dataset called 'bank.csv' and perform association rule mining and sequence pattern mining. 

A bank’s Marketing department is interested in examining associations between various retail banking services used by its customers. The Marketing department would like to determine both typical and atypical service combinations. The dataset is provided in the file `bank.csv`.

These requirements will suffice to conducting association mining on the dataset, a market basket analysis. The
data for this problem consists of two variables: a transaction ID and an item. For each transaction, there is a list of items. For the banking
dataset, a transaction is an individual customer account, and items are products bought by the customer. An association rule is a statement of the form (item set A) => (item set B).

Recall from the lecture that the most common association rule mining algorith is **Apriori algorithm**. Unfortunately, `sklearn` does not provide an implementation of Apriori algorithm. Therefore, we will install and use a
library called `apyori` for this task.

Go to Anaconda prompt and install the library using this command:

```bash
pip install apyori
```

Once the library is installed, we need to perform some data preprocessing on the `bank` dataset. Firstly, load the data set using pandas.

In [1]:
import pandas as pd

# load the bank transaction dataset
df = pd.read_csv('bank.csv')

# info and the first 10 transactions
print(df.info())
print(df.head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32367 entries, 0 to 32366
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   ACCOUNT  32367 non-null  int64 
 1   SERVICE  32367 non-null  object
 2   VISIT    32367 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 758.7+ KB
None
   ACCOUNT SERVICE  VISIT
0   500026   CKING      1
1   500026     SVG      2
2   500026     ATM      3
3   500026     ATM      4
4   500075   CKING      1
5   500075    MMDA      2
6   500075     SVG      3
7   500075     ATM      4
8   500075   TRUST      5
9   500075   TRUST      6


The BANK data set contains service information of nearly 8,000 customers. There are three variables in the data set:
1. ACCOUNT: Account number, nominal
2. SERVICE: Type of service, nominal
3. VISIT: Order of product purchased, ordinal

The dataset has over 32,000 rows. Each row represents a customer-service combination. Therefore, a single customer can have multiple rows in the data set, and each row represents one of the products he or she owns. The median number of products per customer is three. The 13 products are represented in the data set using the following abbreviations:

* ATM - automated teller machine debit card
* AUTO automobile installment loan
* CCRD credit card
* CD certificate of deposit
* CKCRD check/debit card
* CKING checking account
* HMEQLC home equity line of credit
* IRA individual retirement account
* MMDA money market deposit account
* MTG mortgage
* PLOAN personal/consumer installment loan
* SVG saving account
* TRUST personal trust account

As we are looking to generate association rules from items purchased by each account holder, we need to group the accounts and then generate the list of all services purchased.

In [2]:
# group by account, then list all services
transactions = df.groupby(['ACCOUNT'])['SERVICE'].apply(list)

print(transactions.head(5))

ACCOUNT
500026                   [CKING, SVG, ATM, ATM]
500075    [CKING, MMDA, SVG, ATM, TRUST, TRUST]
500129              [CKING, SVG, IRA, ATM, ATM]
500256               [CKING, SVG, CKCRD, CKCRD]
500341               [CKING, SVG, CKCRD, CKCRD]
Name: SERVICE, dtype: object


Once the `transactions` table containing all services purchased by each account number is populated, we are ready to generate association rules. The `apyori`'s `apriori` function accepts a number of parameters, mainly:
1. `transactions`: list of list of items in transactions (eg. [['A', 'B'], ['B', 'C']]).
2. `min_support`: Minimum support of relations in float percentage. It specifies a minimum level of support to claim that items are associated (i.e. they occur together in the dataset). Default 0.1.
3. `min_confidence`: Minimum confidence of relations in float percentage. Default 0.0.
4. `min_lift`: Minimum lift of relations in float percentage. Default 0.0.
5. `max_length`: Max length of the relations. Default None.

Note: Parameters `min_support` and `min_confidence` control the numbers and types of rules generated. There are many heuristics that you can apply to set these numbers. (1) If you are interested in generating associations that involve fairly rare products, you should consider reducing `min_support`. (2) If the items present in the dataset do not show high support, 'min_support' threshold shoudl be set to small value and vice-versa. (3) If you obtain too many rules to be practically useful, you should consider increasing `min_suport` and `min_confidence` as a possible solution

We will run the `apyori` model with the pre-processed transactions and `min_support` of 0.05.

In [3]:
from apyori import apriori

# type cast the transactions from pandas into normal list format and run apriori
transaction_list = list(transactions)
results = list(apriori(transaction_list, min_support=0.05))

# print first 5 rules
print(results[:5])

[RelationRecord(items=frozenset({'ATM'}), support=0.3845576273307471, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'ATM'}), confidence=0.3845576273307471, lift=1.0)]), RelationRecord(items=frozenset({'AUTO'}), support=0.09285446126892755, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'AUTO'}), confidence=0.09285446126892755, lift=1.0)]), RelationRecord(items=frozenset({'CCRD'}), support=0.154799149042673, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'CCRD'}), confidence=0.154799149042673, lift=1.0)]), RelationRecord(items=frozenset({'CD'}), support=0.24527593542735576, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'CD'}), confidence=0.24527593542735576, lift=1.0)]), RelationRecord(items=frozenset({'CKCRD'}), support=0.11300212739331748, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'CKCRD'}), confidence=0.1

The output may look very cluttered. The following function can be used to printing them neatly. We will not go deeper to explain how it works and it is not essential for your learning objective, but we have included some comments to help you out.

In [4]:
def convert_apriori_results_to_pandas_df(results):
    rules = []
    
    for rule_set in results:
        for rule in rule_set.ordered_statistics:
            # items_base = left side of rules, items_add = right side
            # support, confidence and lift for respective rules
            rules.append([','.join(rule.items_base), ','.join(rule.items_add),
                         rule_set.support, rule.confidence, rule.lift]) 
    
    # typecast it to pandas df
    return pd.DataFrame(rules, columns=['Left_side', 'Right_side', 'Support', 
                                        'Confidence', 'Lift']) 

result_df = convert_apriori_results_to_pandas_df(results)

print(result_df.head(20))

   Left_side  Right_side   Support  Confidence      Lift
0                    ATM  0.384558    0.384558  1.000000
1                   AUTO  0.092854    0.092854  1.000000
2                   CCRD  0.154799    0.154799  1.000000
3                     CD  0.245276    0.245276  1.000000
4                  CKCRD  0.113002    0.113002  1.000000
5                  CKING  0.857840    0.857840  1.000000
6                 HMEQLC  0.164685    0.164685  1.000000
7                    IRA  0.108372    0.108372  1.000000
8                   MMDA  0.174446    0.174446  1.000000
9                    MTG  0.074334    0.074334  1.000000
10                   SVG  0.618696    0.618696  1.000000
11                CD,ATM  0.071581    0.071581  1.000000
12       ATM          CD  0.071581    0.186137  0.758889
13        CD         ATM  0.071581    0.291837  0.758889
14             CKING,ATM  0.361907    0.361907  1.000000
15       ATM       CKING  0.361907    0.941100  1.097058
16     CKING         ATM  0.361

The table contains statistics of support, confidence and lift for each of the rules.

Consider the rule A ⇒ B. Recall the following theoretical concepts:
* Support of A ⇒ B is the probability that a customer has both A and B.
* Confidence of A ⇒ B is the probability that a customer has B given that the customer has A.
* Lift of A ⇒ B is a measure of strength of the association. The Lift=2 for the rule A=>B indicates that a customer having A is twice as likely to have B than a customer chosen at random.

In a typical setting, we would like to view the rules by lift value. Sort the rules using the code below:

In [5]:
# sort all acquired rules descending by lift
result_df = result_df.sort_values(by='Lift', ascending=False)
print(result_df.head(10))

         Left_side     Right_side   Support  Confidence      Lift
131          CKCRD     CCRD,CKING  0.055813    0.493909  3.325045
134     CCRD,CKING          CKCRD  0.055813    0.375737  3.325045
33            CCRD          CKCRD  0.055813    0.360550  3.190645
130           CCRD    CKING,CKCRD  0.055813    0.360550  3.190645
135    CKING,CKCRD           CCRD  0.055813    0.493909  3.190645
34           CKCRD           CCRD  0.055813    0.493909  3.190645
203     HMEQLC,SVG      CKING,ATM  0.060944    0.546577  1.510268
198      CKING,ATM     HMEQLC,SVG  0.060944    0.168396  1.510268
196         HMEQLC  CKING,SVG,ATM  0.060944    0.370061  1.489001
205  CKING,SVG,ATM         HMEQLC  0.060944    0.245217  1.489001


The highest lift rule is *checking*, and *credit card* implies *check card*. This is not surprising given that many check cards include credit card logs.

Now, set the `min_support` to 0.01 and list the rules generated.

In [6]:
results2 = list(apriori(transaction_list, min_support=0.05))
result2_df = convert_apriori_results_to_pandas_df(results2)
# sort all acquired rules descending by lift
result2_df = result2_df.sort_values(by='Lift', ascending=False)
print(result2_df.head(10))

         Left_side     Right_side   Support  Confidence      Lift
131          CKCRD     CCRD,CKING  0.055813    0.493909  3.325045
134     CCRD,CKING          CKCRD  0.055813    0.375737  3.325045
33            CCRD          CKCRD  0.055813    0.360550  3.190645
130           CCRD    CKING,CKCRD  0.055813    0.360550  3.190645
135    CKING,CKCRD           CCRD  0.055813    0.493909  3.190645
34           CKCRD           CCRD  0.055813    0.493909  3.190645
203     HMEQLC,SVG      CKING,ATM  0.060944    0.546577  1.510268
198      CKING,ATM     HMEQLC,SVG  0.060944    0.168396  1.510268
196         HMEQLC  CKING,SVG,ATM  0.060944    0.370061  1.489001
205  CKING,SVG,ATM         HMEQLC  0.060944    0.245217  1.489001


## 3. Performing Sequential Rule Mining Using SPMF <a name="sm"></a>

This section introduces you how to use an algorithm implemnetd in another lanaguge. There exists no mature library in Python that implements sequential rule mining. On the other hand, SPMF is a Java library implemented by Philippe Fournier Viger containing 150 pattern mining algorithms, such as frequent pattern mining, association rules, frequent sequences and of course, sequential rule mining. We will use SPMF to perform sequential rule mining on this dataset.

More information about SPMF can be found here http://www.philippe-fournier-viger.com/spmf/index.php and you can download SPMF here http://www.philippe-fournier-viger.com/spmf/index.php?link=download.php. Put the `spmf.jar` file into the same directory with this Jupyter notebook.

As SPMF is implemented in Java, you need to install Java to run the `.jar` file. There exist many tutorials on the Internet. To check if you have installed Java correctly for this practical, open your terminal/CMD and type `java`.

To demonstrate how to use SPMF through Python, we will use the same `bank.csv` dataset.

Different from association mining, to get sequential rules, there must be information on the order of the target. In this dataset, the order is given in `VISIT` column. As this dataset is already ordered by `VISIT`, a simple group and apply list will produce sequences in order. In other datasets, you may have to perform some preprocessing to ensure this ordering is correct.

In [7]:
transactions = df.groupby(['ACCOUNT'])['SERVICE'].apply(list)
sequences = transactions.values.tolist()

# show the first 5 sequences
print(sequences[:5])

[['CKING', 'SVG', 'ATM', 'ATM'], ['CKING', 'MMDA', 'SVG', 'ATM', 'TRUST', 'TRUST'], ['CKING', 'SVG', 'IRA', 'ATM', 'ATM'], ['CKING', 'SVG', 'CKCRD', 'CKCRD'], ['CKING', 'SVG', 'CKCRD', 'CKCRD']]


Once you have sequences ordered correctly, you could simply run the following function to get sequential rules and their respective support and confidence. In general, the function below will write the sequences into a file called `seq_rule_input.txt` in SPMF accepted format, run SPMF to generate sequential rules, read the output file and return a Pandas dataframe. Detailed comments is provided in the function source code.

In [8]:
from collections import defaultdict
import subprocess
import re

''' Uses SPMF to find association rules in supplied transactions '''
def get_association_rules(sequences, min_sup, min_conf):
    # step 1: create required input for SPMF
    
    # prepare a dict to uniquely assign each item in the transactions to an int ID
    item_dict = defaultdict(int)
    output_dict = defaultdict(str)
    item_id = 1
    
    # write your sequences in SPMF format
    with open('seq_rule_input.txt', 'w+') as f:
        for sequence in sequences:
            z = []
            for itemset in sequence:
                # if there are multiple items in one itemset
                if isinstance(itemset, list):
                    for item in itemset:
                        if item not in item_dict:
                            item_dict[item] = item_id
                            item_id += 1

                        z.append(item_dict[item])
                else:
                    if itemset not in item_dict:
                        item_dict[itemset] = item_id
                        output_dict[str(item_id)] = itemset
                        item_id += 1
                    z.append(item_dict[itemset])
                    
                # end of itemset
                z.append(-1)
            
            # end of a sequence
            z.append(-2)
            f.write(' '.join([str(x) for x in z]))
            f.write('\n')
    
    # run SPMF with supplied parameters
    supp_param = '{}%'.format(int(min_sup * 100))
    conf_param = '{}%'.format(int(min_conf * 100))
    subprocess.call(['java', '-jar', 'spmf.jar', 'run', 'RuleGrowth', 
                     'seq_rule_input.txt', 'seq_rule_output.txt', 
                     supp_param, conf_param], shell=True)
    
    # read back the output rules
    outputs = open('seq_rule_output.txt', 'r').read().strip().split('\n')
    output_rules = []
    for rule in outputs:
        left, right, sup, conf = re.search(pattern=r'([0-9\,]+) ==> ([0-9\,]+) #SUP: ([0-9]+) #CONF: ([0-9\.]+)', string=rule).groups()
        sup = int(sup) / len(sequences)
        conf = float(conf)
        output_rules.append([[output_dict[x] for x in left.split(',')], [output_dict[x] for x in right.split(',')], sup, conf])
    
    # return pandas DataFrame
    return pd.DataFrame(output_rules, columns = ['Left_rule', 'Right_rule', 'Support', 'Confidence'])

You can run the function on sequences from the `bank.csv` dataset using the command below. In here, we are using `min_supp` of 0.1 and `min_conf` of 0.1.

In [9]:
get_association_rules(sequences, 0.1, 0.1)

Left_rule     Right_rule   Support  Confidence
0        [CKING]          [SVG]  0.541734    0.631510
1        [CKING]     [SVG, ATM]  0.248530    0.289716
2        [CKING]      [SVG, CD]  0.142535    0.166156
3        [CKING]  [SVG, HMEQLC]  0.111500    0.129978
4        [CKING]          [ATM]  0.361907    0.421882
5   [CKING, SVG]          [ATM]  0.248530    0.458766
6        [CKING]         [MMDA]  0.155800    0.181619
7        [CKING]        [CKCRD]  0.113002    0.131729
8        [CKING]           [CD]  0.209861    0.244639
9   [CKING, SVG]           [CD]  0.142535    0.263109
10       [CKING]         [CCRD]  0.148542    0.173158
11       [CKING]       [HMEQLC]  0.164685    0.191977
12  [CKING, SVG]       [HMEQLC]  0.111500    0.205821
13         [SVG]          [ATM]  0.256914    0.415251
14         [SVG]           [CD]  0.157177    0.254045
15         [SVG]         [CCRD]  0.102240    0.165251
16         [SVG]       [HMEQLC]  0.111500    0.180218

The first rule is `CKING` => `SVG` with 0.54 support and 0.63 confidence. This is a strong rule. The support value implies that 54% of customers bought `SVG` after `CKING`. The confidence value implies that if a customer has bought `CKING`, the probability of them buying `SVG` subsequently is 63%.

## End Notes

We learned how to build, tune and explore association mining models. We also used visualisation to help us explain the association rules. The goal of association mining is to identify association among varaiables without the presence of target variable explictly and training of the model. This analysis is based on frequency of items present in the transactional dataset.